---
title: "Gila Monster Sex Chromosomes"
author: "Annika Vannan"
date: "01/25/2022"
output:
  rmdformats::downcute:
  code_folding: show
self-contained: TRUE
thumbnails: FALSE
lightbox: TRUE
pkgdown:
  as_is: TRUE
params:
  work_dir: "G:/Noctsol/R Projects/GilaMonster/Ancestral_1_24_2022/"
  filename: "z_ortho_filtered.corrected.stringtie_compiled_per_transcript_separate_individuals.gila2_galgal5.galgal5_refbased.txt"
  chicken_file: "gila2_galgal5_gff_comparison.txt"
  feature_type: "transcript"
---
  
```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = params$show_code, error = TRUE)
```

# Set-up
## Environment
Before running this script, you may need to change the parameters of this R 
Markdown file.

The parameters include:
  
  * **work_dir:** *Your working directory, which contains the files to be loaded*
  * **filename:** *Tab-delimited counts file, with counts per transcript/exon*
  * **chicken_file:** *File for comparing Gila to chicken; "gila2_galgal5_gff_comparison.txt"*
  * **feature_type:** *String; either "transcript" or "exon"*
  
Once the parameters have been set, load the necessary libraries and file. We 
will want to run this script with 2 different files:
  1. "z_ortho_filtered.corrected.stringtie_compiled_per_exon_separate_individuals.gila2_galgal5.galgal5_refbased.txt"
  2. "z_ortho_filtered.corrected.stringtie_compiled_per_transcript_separate_individuals.gila2_galgal5.galgal5_refbased.txt"

Note that throughout the code, "transcripts" and "exons" might be used 
interchangeably - check the file you loaded in to verify whether you are 
actually looking at exons or transcripts. Change the parameter "feature_type"
to "exon" or "transcript" accoringly.

```{r load, message = FALSE, warning = FALSE}
# Libraries
library(tidyverse)
library(rstatix)
library(knitr)
library(ggplot2)
library(ggpubr)
library(ggrepel)
library(ggbeeswarm)
library(clipr)

# Exon/transcript file
compiled_counts <- read_tsv(paste0(params$work_dir, params$filename))
glimpse(compiled_counts)

# Chicken comparison file
# Unite GFF2 scaffold and start columns for easier filtering
gal_comp <- read_tsv(paste0(params$work_dir, params$chicken_file)) %>%
  unite("scaff_start", c(GFF2_scaff, GFF2_start), sep = "_", remove = FALSE)

# Scientific notation is a pain
options(scipen = 99999)
```

## Tidy the Data
In tidying the data, we will keep only the scaffolds we are interested in 
comparing. The data will then converted from wide to long format. The PAR will
also be removed - it is located on scaffold 304 and spans from the start until 
position 1,740,000. Most of these transcripts/exons do not exist in our chicken 
comparison file, but some do, so we will need to remove them.

```{r, tidy}
# Get PAR position and all other sex scaffold positions
remove_par <- gal_comp %>% 
  filter(type == params$feature_type, GFF1_scaff == "304" & GFF1_start < 1740000) %>%
  pull(scaff_start)
sex_scaff_start <- gal_comp %>% 
  filter(type == params$feature_type, !(GFF1_scaff == "304" & GFF1_start < 1740000)) %>%
  pull(scaff_start)

length(remove_par) # 14 transcripts, 149 exons

counts <- compiled_counts %>%
  # This select function should be run for "corrected"
  select(c(1, 2, 3), ncol(.) - c(1, 2)) %>%
  # This select function should be run for uncorrected
#  select(1:5) %>%
  unite("scaff_start", c(scaffold, start), sep = "_", remove = FALSE) %>%
  mutate(chrom_type = 
           case_when(scaff_start %in% sex_scaff_start ~ "Sex", 
                     (scaffold %in% c("chr1", "chr2", "chr3", "chr4")) ~ "Auto",
                     TRUE ~ "None")) %>%
  filter(!chrom_type == "None") %>%
  pivot_longer(cols = contains("_mean"),
               names_to = "sex", values_to = "Mean") %>%
  mutate(sex = case_when(str_detect(sex, "female") ~ "Female", 
                         str_detect(sex, "male") ~ "Male"))

glimpse(counts)
```

# Statistics
## Filter Transcripts/Exons
Expression values for transcripts/exons are in FPKM. Here, we will remove
transcripts/exons that are unexpressed (0 FPKM) in **BOTH** sexes. If they are
unexpressed in one sex but expressed in the other, they will be retained.

```{r, filter, message = FALSE}
# Transcripts BEFORE filtering, stratified by sex and chromosome
before_filter <- counts %>%
  select(chrom_type, contains("id")) %>%
  group_by(chrom_type) %>%
  summarize(transcripts_before = n()/2) %>%
  unique()

# Filter based on expression
counts_filtered_par <- counts %>%
  setNames(c("id", "scaff_start", "scaffold", "start", 
             "chrom_type", "sex", "Mean")) %>%
  group_by(id) %>%
  mutate(sum = sum(Mean)) %>%
  filter(sum > 0) %>%
  select(-sum) %>%
  ungroup()

counts_filtered_nopar <- counts_filtered_par %>%
  # Remove PAR
  filter(!scaff_start %in% remove_par)

counts_filtered <- counts_filtered_nopar %>%
  filter(chrom_type == "Sex" | scaffold == "chr2")

# Transcripts AFTER filtering, stratified by sex and chromosome type
after_filter <- counts_filtered %>%
  select(chrom_type, sex, contains("id")) %>%
  group_by(chrom_type, sex) %>%
  summarize(transcripts_after = n()) %>%
  select(chrom_type, transcripts_after) %>%
  unique()

# Merge before & after counts into a summary table
transcript_counts <- inner_join(before_filter, after_filter)

transcript_counts
```

```{r, filter_hidden, include = FALSE}
# General summary data (not stratified by sex or chromosome)
num_bf <- nrow(counts)
num_af <- nrow(counts_filtered)
percent_left <- round((num_af/num_bf) * 100, 1)
```

Out of `r num_bf` transcripts, `r num_af` (`r percent_left`%) are left after
filtering. 

## Expression Ratios
To assess possible dosage compensation, overall transcript expression on the
putative Z chromosome can be compared to expression on the autosomes for each
sex. We will get the following ratios:
  
  * **Female_ZA:** *(Female Z / Female Auto)*
  * **Male_ZA:** *(Male Z / Male Auto)*
  * **Auto_FM:** *(Female Auto / Male Auto)*
  * **Sex_FM** *(Female Z / Male Z)*
  
The original counts file contains mean transcript expression per sex (n=3/sex).
Here, we will calculate the expression ratios based on either the the means or
the medians of these means. Those values will be contained in the `mean_mean_all` 
and `median_mean_all` columns in a new dataframe, `full_mean_median`.

```{r, ratios, message = FALSE}
# Means and medians of filtered counts - by sex and chromosome
full_mean_median <- counts_filtered %>%
  group_by(sex, chrom_type) %>%
  summarize(mean_mean_all = mean(Mean), 
            median_mean_all = median(Mean)) %>%
  arrange(chrom_type)

# Means and medians of filtered counts - by sex and scaffold (Z only)
z_scaffold_summary <- counts_filtered %>%
  filter(chrom_type == "Sex") %>%
  group_by(sex, chrom_type, scaffold) %>%
  summarize(mean_mean_all = mean(Mean), 
            median_mean_all = median(Mean))

# Means and medians of filtered counts - by sex and scaffold (Auto only)
auto_scaffold_summary <- counts_filtered %>%
  filter(chrom_type == "Auto") %>%
  group_by(sex, chrom_type, scaffold) %>%
  summarize(mean_mean_all = mean(Mean), 
            median_mean_all = median(Mean))

# Expression ratios: Means
mean_ratios <- full_mean_median %>%
  select(-contains("median")) %>%
  pivot_wider(names_from = c("sex", "chrom_type"), 
              values_from = c("mean_mean_all")) %>%
  transmute(Auto_FM = Female_Auto/Male_Auto, 
            Sex_FM = Female_Sex/Male_Sex, 
            Female_ZA = Female_Sex/Female_Auto, 
            Male_ZA = Male_Sex/Male_Auto)

# Expression ratios: Medians
median_ratios <- full_mean_median %>%
  select(-contains("mean_mean")) %>%
  pivot_wider(names_from = c("sex", "chrom_type"), 
              values_from = c("median_mean_all")) %>%
  transmute(Auto_FM = Female_Auto/Male_Auto, 
            Sex_FM = Female_Sex/Male_Sex, 
            Female_ZA = Female_Sex/Female_Auto, 
            Male_ZA = Male_Sex/Male_Auto)
```

Create a plot that shows the expression compared between Sex and Chromosome Type.
This plot is somewhat inaccurate because it groups by scaffold.

```{r, ratio_plot, message = FALSE, fig.width = 4, fig.height = 4}
ratio_plot <- rbind(auto_scaffold_summary, z_scaffold_summary) %>%
  ggplot(aes(x = sex, y = mean_mean_all, fill = chrom_type, color = chrom_type)) +
  geom_boxplot(width = 0.5, color = "black") +
  geom_point(position = position_dodge(width = 0.5), size = 2.5) +
  scale_color_manual(name = "Chromosome", labels = c("Autosome", "Z"),,
                     values = c("grey10", "276fd5")) +
  scale_fill_manual(name = "Chromosome", labels = c("Autosome", "Z"),
                    values = c("grey70", "#023785")) +
  labs(x = "Sex", y = "Mean Expression (FPKM)") +
  theme_classic(base_size = 14) +
  theme(legend.position = "bottom", legend.key.size = unit(0.75, "cm"),
        axis.text = element_text(color = "black"), axis.ticks.x = element_blank(),
        legend.title.align = 0.5, plot.margin = unit(c(0.5, 0.5, 0.5, 0.5), "cm"))

ratio_plot
```

We can also visualize the distributions more precisely. Adjust the y axis limits
as-needed using coord_cartesian()

```{r, swarm, fig.width = 4.5, fig.height = 4.5}
swarm_plot <- counts %>%
  ggplot(aes(x = sex, y = Mean, fill = chrom_type)) +
#  coord_cartesian(ylim = c(0, 10)) +
  geom_quasirandom(aes(group = interaction(sex, chrom_type)), dodge.width = 0.8, 
                   shape = 21, alpha = 0.8, cex = 2.5, nbins = 15) +
  geom_boxplot(aes(group = interaction(sex, chrom_type)), 
               position = position_dodge(width = 0.8),
               fill = "white", width = 0.2, color = "black", 
               outlier.color = NA) +
  scale_fill_manual(name = "Chromosome", labels = c("Autosome", "Z"),
                    values = c("grey60", "red")) +
  theme_classic(base_size = 13) +
  theme(legend.position = "bottom", legend.key.size = unit(0.75, "cm"),
        axis.text = element_text(color = "black"), 
        axis.ticks.x = element_blank(), legend.title = element_blank()) +
  labs(x = "", y = "Distribution of Exon Means")

swarm_plot
```

## Comparing Expression
The `rstatix` package comes with handy functions that allow you to perform
statistical tests in tandem with `tidyverse` commands. We will use it to compare
the same groups as above with the expression ratios using the nonparametric
Wilcoxon rank-sum tests to compare means between groups.

```{r, wilcox}
# Auto F vs. M, paired
wilcox_results1 <- counts_filtered %>%
  filter(chrom_type == "Auto") %>%
  wilcox_test(Mean ~ sex, p.adjust = "none", paired = TRUE) %>%
  mutate(comparison = "Auto_FM", .before = everything()) %>%
  mutate(sig = case_when(p < 0.05 ~ "*", TRUE ~ "ns")) %>%
  select(comparison, statistic, p, sig)

test <- counts_filtered %>%
  filter(chrom_type == "Auto") %>%
  pivot_wider(names_from = sex, values_from = Mean)
  
wilcox.test(test$Female, test$Male, paired = TRUE)


# Z F vs. M, paired
wilcox_results2 <- counts_filtered %>%
  filter(chrom_type == "Sex") %>%
  wilcox_test(Mean ~ sex, p.adjust = "none", paired = TRUE) %>%
  mutate(comparison = "Sex_FM", .before = everything()) %>%
  mutate(sig = case_when(p < 0.05 ~ "*", TRUE ~ "ns")) %>%
  select(comparison, statistic, p, sig)

# Not paired
wilcox_results3 <- counts_filtered %>%
  unite(col = "sex_chromtype", c("sex", "chrom_type"), sep = "") %>%
  wilcox_test(Mean ~ sex_chromtype, p.adjust = "none") %>%
  filter(!(group1 == "FemaleAuto" & group2 == "MaleSex"),
         !(group1 == "FemaleSex" & group2 == "MaleAuto")) %>%
  unite(col = "comparison", c(group1, group2), sep = "", remove = TRUE) %>%
  mutate(comparison = recode_factor(comparison, 
                                    `FemaleAutoMaleAuto` = "Auto_FM",
                                    `FemaleSexMaleSex` = "Sex_FM",
                                    `FemaleAutoFemaleSex` = "Female_ZA",
                                    `MaleAutoMaleSex` = "Male_ZA",
                                    .ordered = TRUE)) %>%
#  filter(comparison == "Female_ZA" | comparison == "Male_ZA") %>%
  arrange(comparison) %>%
  select(-c(n1, n2, p.adj, .y.)) %>%
  rename(sig = p.adj.signif)

#wilcox_results <- rbind(wilcox_results1, wilcox_results2, wilcox_results3)

wilcox_results <- wilcox_results3
```

It can be difficult to interpret the results of Wilcoxon rank-sum tests when the
means and medians are very different. Instead, we can calculate the actual
mean ranks for each comparison, and use that to determine which is "higher" or
"lower" in expression within a comparison.

```{r, mean_ranks}
# Female Z:A
counts_filtered %>%
  unite(col = "sex_chromtype", c("sex", "chrom_type"), sep = "") %>%
  arrange(Mean) %>%
  filter(sex_chromtype == "FemaleAuto" | sex_chromtype == "FemaleSex") %>%
  mutate(Rank = rank(Mean)) %>%
  group_by(sex_chromtype) %>%
  summarize(Mean_Rank = mean(Rank))

# Male Z:A
counts_filtered %>%
  unite(col = "sex_chromtype", c("sex", "chrom_type"), sep = "") %>%
  arrange(Mean) %>%
  filter(sex_chromtype == "MaleAuto" | sex_chromtype == "MaleSex") %>%
  mutate(Rank = rank(Mean)) %>%
  group_by(sex_chromtype) %>%
  summarize(Mean_Rank = mean(Rank))

# M:F Autosome
counts_filtered %>%
  unite(col = "sex_chromtype", c("sex", "chrom_type"), sep = "") %>%
  arrange(Mean) %>%
  filter(sex_chromtype == "MaleAuto" | sex_chromtype == "FemaleAuto") %>%
  mutate(Rank = rank(Mean)) %>%
  group_by(sex_chromtype) %>%
  summarize(Mean_Rank = mean(Rank))

# M:F Z chromosome
counts_filtered %>%
  unite(col = "sex_chromtype", c("sex", "chrom_type"), sep = "") %>%
  arrange(Mean) %>%
  filter(sex_chromtype == "MaleSex" | sex_chromtype == "FemaleSex") %>%
  mutate(Rank = rank(Mean)) %>%
  group_by(sex_chromtype) %>%
  summarize(Mean_Rank = mean(Rank))
```



### ADDED 01/27/2022
```{r, log_z, echo = FALSE, fig.height = 8, fig.width = 5}
counts_filtered %>%
  pivot_wider(names_from = sex, values_from = Mean) %>%
  mutate(f_m_ratio = Female/Male, log2_f_m = log2(f_m_ratio)) %>%
  filter(chrom_type == "Sex", log2_f_m != Inf, log2_f_m != -Inf) %>%
  ggplot(aes(x = start, y = log2_f_m)) +
  geom_point() +
  geom_smooth() +
  # geom_hline(yintercept = c(0, -1, 1), color = "black", 
  #            linetype = rep(c("solid", "dashed", "dashed"), length(6))) +
  labs(x = "Transcript Start", y = expression("log"[2]*" F:M Ratio (Means)"), title = "Z Chromosome") +
  theme_bw() +
  theme(axis.text.x = element_text(vjust = 0.5), axis.text = element_text(color = "black"),
        plot.title = element_text(hjust = 0.5)) +
  facet_wrap(~ scaffold, scales = "free")
```

# Complete Statistical Results!
All of the results are combined into one list called `full results`, but in an R
Markdown document, it's cleaner to view them individually.

```{r, echo = FALSE}
full_results <- list(transcript_counts = transcript_counts, 
                     summary_data = full_mean_median,
                     z_scaffold_summary = z_scaffold_summary,
                     auto_scaffold_summary = auto_scaffold_summary,
                     wilcox_results = wilcox_results,
                     mean_ratios = mean_ratios,
                     median_ratios = median_ratios)
```

```{r, individual_results, results = "asis"}
kable(transcript_counts)
kable(full_mean_median)
kable(z_scaffold_summary)
kable(auto_scaffold_summary)
kable(wilcox_results)
kable(mean_ratios)
kable(median_ratios)
```


We can also create a plot visualizing the means of the mean expression values
for each sex and chromosome type. Adjust the y axis limits as necessary using
scale_y_continuous().

```{r, echo = FALSE, message = FALSE, fig.width = 4.5, fig.height = 4.5}
full_mean_median %>%
  mutate(chrom_type = ifelse(chrom_type == "Sex", "Z", "Auto")) %>%
  ggplot(aes(x = sex, y = median_mean_all, fill = chrom_type)) +
  geom_point(position = position_dodge(width = 0.8), size = 15, shape = 22, 
             stroke = 1) +
  scale_fill_manual(values = c("grey60", "red")) +
  labs(x = "", y = "Corrected FPKM, Median of Means") +
  theme_classic(base_size = 13) +
  theme(legend.position = "bottom", legend.key.size = unit(0.75, "cm"),
        axis.text = element_text(color = "black"), 
        axis.ticks.x = element_blank(), legend.title = element_blank())

full_mean_median %>%
  mutate(chrom_type = ifelse(chrom_type == "Sex", "Z", "Auto")) %>%
  ggplot(aes(x = sex, y = mean_mean_all, fill = chrom_type)) +
  geom_point(position = position_dodge(width = 0.8), size = 15, shape = 22, 
             stroke = 1) +
  scale_y_continuous(limits = c(0, 3)) +
  scale_fill_manual(values = c("grey60", "red")) +
  labs(x = "", y = "Corrected FPKM, Mean of Means") +
  theme_classic(base_size = 13) +
  theme(legend.position = "bottom", legend.key.size = unit(0.75, "cm"),
        axis.text = element_text(color = "black"), 
        axis.ticks.x = element_blank(), legend.title = element_blank())
```